### PR TITLE
Create Role 0: Schema

### PR DESCRIPTION
**Pull request by KH-Amir-TN**
## Description

Enable users with the right set of privileges to create roles.

This PR completes **0.** 
--
~~0. Add `color:string` attribute to roles model, `null` values aren't allowed and should default to the app primary HEX color.~~
1. Add `RolesController#create` to create a new role while providing its name.
> When permissions is made Roles just like `RoomMeetingOptions` should take the default
configuration on `:create` and can be updated later on.
2. Extend the UI and create a CreateRole modal that prompts  the user for a role name when clicking a button in the Roles page of the organization settings.
3. Extend/Add required mutations for integration with the `RolesController#create` API and
show proper feedbacks.
 
### User story [Manage Roles]:
---
1. A User with the right set of priviledges authenticates.
2. User goes to the Organasation settings.
3. User goes to the Roles page.
4. User can select a role and:
    - Delete it (Not all roles can be deleted).
[OR]
    - Edit color, name, permissions.
[OR]
    - Add new role while providing its name and maybe color.

## Testing Steps

> 1. Pull the code.
> 2. Install the dependencies `bundle install && npm|yarn install`.
> 3. Clean the previous assets build by running `rm app/assets/builds/*` (This won't remove .keep since it's hidden).
> 4. Clean the database and tmp files for a better isolation by running `rails tmp:clear && rails db:schema:cache:clear && rails db:drop && rails db:create && rails db:migrate:with_data`
> 5. Run the linter and specs `bundle exec rubocop --parallel && bundle exec rspec && npx eslint app/javascript/* --ext .jsx,.js`
> 6. Run `./bin/dev` to run the assets builders processes and the Puma server all at once.

### CODE REVIEW

**Review by scouillard — APPROVED** [review]

**Comment by farhatahmad** [comment]
I'm thinking of partially changing this requirement. Role color has little impact and allowing the admin to change the colors doesn't really seem that useful. 

I think we should set some default colors for the 3 base roles and then just auto generate a new color for any new role that's created

File: app/validators/hex_color_validator.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# frozen_string_literal: true
+
+class HexColorValidator < ActiveModel::EachValidator
+  HEX_COLOR_PATTERN = /\A#([A-F0-9]{6}|[A-F0-9]{3})\Z/i
+
+  def validate_each(record, attribute, value)
+    record.errors.add(attribute, message: 'Invalid CSS hex color code.') unless HEX_COLOR_PATTERN.match?(value) # TODO: Add I18n.
+  end
+end
```
Comment (farhatahmad) [inline app/validators/hex_color_validator.rb:9]: I think we can drop this

File: app/models/role.rb [hunk 2]
```diff
@@ -4,4 +4,13 @@ class Role < ApplicationRecord
...
 
   validates :name, presence: true
+  validates :color, hex_color: true
+
+  before_validation :set_default_color, on: :create
+
+  private
+
+  def set_default_color
+    self.color = Rails.configuration.styles[:primary] unless color?
```
[the remark this answers is not in the document]
Comment (KH-Amir-TN) [reply app/models/role.rb:14]: Likes the idea,

File: app/validators/hex_color_validator.rb (hunk 1 above)
Comment (KH-Amir-TN) [reply app/validators/hex_color_validator.rb:9]: Yup, if we're generating the colors

File: app/models/role.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (KH-Amir-TN) [reply app/models/role.rb:14]: For the update I think we should give the admins the right to generate a new random color from the client app just like with GitHub labels?

File: app/models/role.rb [hunk 3]
```diff
@@ -4,4 +4,12 @@ class Role < ApplicationRecord
   has_many :users, dependent: :restrict_with_exception
 
   validates :name, presence: true
+
+  before_validation :set_random_color, on: :create
+
+  private
+
+  def set_random_color
+    self.color = "##{SecureRandom.hex(3)}"
```
Comment (farhatahmad) [inline app/models/role.rb:13]: To avoid collisions, isn't 6 char hex better?

File: db/migrate/20220531180853_create_roles.rb [hunk 4]
```diff
@@ -4,6 +4,7 @@ class CreateRoles < ActiveRecord::Migration[7.0]
   def change
     create_table :roles do |t|
       t.string :name, null: false
+      t.string :color, default: '', null: false
```
Comment (farhatahmad) [inline db/migrate/20220531180853_create_roles.rb:7]: default: '' allows empty records in the db which isn't what we want

File: app/models/role.rb (hunk 3 above)
Comment (KH-Amir-TN) [reply app/models/role.rb:13]: What do you mean?
`SecureRandom.hex(3)` generates 6 hex chars random.
It's a bit deceiving but the parameter isn't the length but the byteLength.

File: db/migrate/20220531180853_create_roles.rb (hunk 4 above)
Comment (KH-Amir-TN) [reply db/migrate/20220531180853_create_roles.rb:7]: The problem was with rubocop, I haven't thought about it.
Should I disable the convention for that line?

Comment (KH-Amir-TN) [reply db/migrate/20220531180853_create_roles.rb:7]: Have you faced something like this when setting `null: false`?

File: app/models/role.rb (hunk 3 above)
Comment (farhatahmad) [reply app/models/role.rb:13]: :facepalm:

File: db/migrate/20220531180853_create_roles.rb (hunk 4 above)
Comment (farhatahmad) [reply db/migrate/20220531180853_create_roles.rb:7]: Hmm that should be fine then

**Review by farhatahmad — APPROVED** [review]

**Pull request merged**